### PR TITLE
HAAR-1371: Auto assign roles while adding groups

### PR DESCRIPTION
**Pull request by ravmoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/repository/RoleRepositoryTest.kt [hunk 1]
```diff
@@ -49,6 +49,27 @@ class RoleRepositoryTest {
...
+      assertThat(repository.findAutomaticGroupRolesByGroupCode("SITE_1_GROUP_2").toList()).size()
+        .isEqualTo(2)
+    }
+
+    @Test
+    fun givenAnNonAutomaticGroupRoleTheyCantBeRetrieved(): Unit = runBlocking {
+      // SITE_1_GROUP_1 have 2 automatic set to 'true' and 1 set to 'false'
+      assertThat(repository.findAutomaticGroupRolesByGroupCode("SITE_1_GROUP_1").toList()).size()
+        .isEqualTo(2)
+    }
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/repository/RoleRepositoryTest.kt:65]: Do we need a test with a role that doesn't have any assigned groups?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupService.kt [hunk 2]
```diff
@@ -96,9 +102,17 @@ class UserGroupService(
...
 
-      roleRepository.saveAll(roleRepository.findRolesByGroupCode(group.groupCode).toList())
+      val groupRoles = roleRepository.findAutomaticGroupRolesByGroupCode(groupCode).toSet()
+
+      val existingUserRoles = userRoleService.getUserRoles(userId)
+
+      // remove existing roles to the user
+      val roleCodesToAdd = existingUserRoles?.let { groupRoles.subtract(it.toSet()) }!!.map { it.roleCode }
+
+      userRoleService.addRolesByUserId(userId, roleCodesToAdd)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupService.kt:112]: I think we did discuss this but wondering if the "distinct" or "distinctBy{it.roleCode} might be nicer/simpler

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/repository/RoleRepositoryTest.kt (hunk 1 above)
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/repository/RoleRepositoryTest.kt:65]: added test covering this scenario

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**